### PR TITLE
Add custom component docs

### LINKED ISSUE #131: Create 'Creating a custom component' documentation

### PR DESCRIPTION
**Pull request by NielsRogge**
Fixes #131

### CODE REVIEW

File: docs/custom_component.md [hunk 1]
```diff
@@ -0,0 +1,81 @@
+# Custom component
+
+The goal of Fondant is to make it easy for people to define new components which can be incorporated into data preparation pipelines. This allows to share and reuse them as building blocks when preparing data for Foundation models.
+
+A Fondant component should implement one logical piece of data preprocessing, like captioning images or removing PPI from text. It's important to not include too much logic into a single component, as it might be beneficial to split up the logic into 2 separate components in that case.
```
Comment (GeorgesLorre) [inline docs/custom_component.md:5]: I had to look up `PPI` but I think it should be `PII` either way I would write it out?

File: docs/custom_component.md [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+To implement a custom component, a couple of files need to be defined:
+
+- [Fondant component specification](#fondant-component-specification)
+- [main.py script in a `src` folder](#mainpy-script)
```
Comment (RobbeSneyders) [inline docs/custom_component.md:10]:
```suggestion
- [`main.py` script in a `src` folder](#mainpy-script)
```

File: docs/custom_component.md [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+- [Dockerfile](#dockerfile)
+- [requirements.txt](#requirementstxt)
+
+To see an overview of all custom components which Fondant already provides, have a look [here](https://github.com/ml6team/fondant/tree/main/components).
```
Comment (RobbeSneyders) [inline docs/custom_component.md:14]:
```suggestion
To see an overview of all reusable components which Fondant already provides, have a look [here](https://github.com/ml6team/fondant/tree/main/components).
```

File: docs/custom_component.md [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+        
+        Args:
+            dataframe: A Dask dataframe containing the data
+            argumentX: A custom argument passed to the component
+        """
+```
+
+The idea is that Fondant provides a single dataframe to the user based on the component's specification. The column names of the dataframe always have the format `subset_field`. This means that if a component defines the `images` subset with the `data` field in its `consumes` [section](component_spec) for instance, the dataframe will contain a column called `images_data`.
+
+Next, the user can manipulate this dataframe. Finally, the user should return a single dataframe, which Fondant will then use to update the [manifest](manifest) and potentially write new data to the cloud. Fondant will also verify that the output dataframe matches with the `produces` [section](component_spec) of the component specificaton.
```
Comment (RobbeSneyders) [inline docs/custom_component.md:40]: Would also explicitly mention that we expect the columns of the returned dataframe in the `subset_field` format.

File: docs/custom_component.md [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+
+Note that the `transform` method can include additional custom arguments (`argument1` and `argument2` in the example above). These should match with the `args` section defined in the [Fondant specification](component_spec). Examples include the batch size when captioning images, the minimum width and height when filtering images, and so on.
+
+Note that the `main.py` script can be split up into several Python scripts in case it would become prohibitively long. See the [prompt based LAION retrieval component](https://github.com/ml6team/fondant/tree/main/components/prompt_based_laion_retrieval/src) as an example: the CLIP client itself is defined in a separate script called `clip_client`, which is then imported in the `main.py` script.
+
+## Dockerfile
+
+A Dockerfile defines all the commands a user could call on the command line to assemble a Docker image. Docker uses this file to build the Docker image automatically by reading the instructions. An example Dockerfile is defined below.
```
Comment (RobbeSneyders) [inline docs/custom_component.md:48]:
```suggestion
The `Dockerfile` defines how to build the component into a Docker image. An example Dockerfile is defined below.
```

File: mkdocs.yml [hunk 6]
```diff
@@ -30,6 +30,7 @@ nav:
   - Home: index.md
   - Manifest: manifest.md
   - Component spec: component_spec.md
+  - Custom component: custom_component.md
```
Comment (RobbeSneyders) [inline mkdocs.yml:33]:
```suggestion
  - Creating custom components: custom_component.md
```

File: docs/custom_component.md [hunk 7]
```diff
@@ -0,0 +1,81 @@
+# Custom component
```
Comment (RobbeSneyders) [inline docs/custom_component.md:1]:
```suggestion
# Creating custom components
```

**Review by RobbeSneyders — COMMENTED** [review]
Thanks @NielsRogge!

Looks great, left some minor comments.

File: docs/component_spec.md [hunk 8]
```diff
@@ -56,7 +56,7 @@ consumes:
   captions:
     fields:
       text:
-        type: string
+        type: utf8
```
[the remark this answers is not in the document]
Comment (NielsRogge) [reply docs/component_spec.md:59]: Can you point me to the PR that fixes this?

Looks like it's still utf8: https://github.com/ml6team/fondant/blob/3f99e062419a6346d5b0e014c3ddc03d29566384/components/caption_images/fondant_component.yaml#L15

**Review by RobbeSneyders — APPROVED** [review]
Thanks Niels!

**Pull request merged**